solana/rpc/providers/ws.py

```diff
@@ -0,0 +1,131 @@
+"""WebSocket RPC provider."""
+# https://github.com/ethereum/web3.py/blob/d431db5fcaa3d624e8361a99ae3317b82367fe65/web3/providers/websocket.py#L83
+import asyncio
+import itertools
+import json
+import logging
+import os
+from threading import Thread
+from types import TracebackType
+from typing import Any, Optional, Type, Union, cast
+
+import websockets
+
+from .._utils.encoding import FriendlyJsonSerde
+from ..types import URI, RPCMethod, RPCResponse
+from .base import BaseProvider
+
+RESTRICTED_WEBSOCKET_KWARGS = {"uri", "loop"}
+DEFAULT_WEBSOCKET_TIMEOUT = 10
+
+
+def get_default_endpoint() -> URI:
+    """Get the default ws rpc endpoint."""
+    return URI(os.environ.get("SOLANA_WS_PROVIDER_URI", "ws://localhost:8900"))
+
+
+def _start_event_loop(loop: asyncio.AbstractEventLoop) -> None:
+    """Start event loop."""
+    asyncio.set_event_loop(loop)
+    loop.run_forever()
+    loop.close()
+
+
+def _get_threaded_loop() -> asyncio.AbstractEventLoop:
+    """Get threaded loop."""
+    new_loop = asyncio.new_event_loop()
+    thread_loop = Thread(target=_start_event_loop, args=(new_loop,), daemon=True)
+    thread_loop.start()
+    return new_loop
+
+
+class PersistentWebSocket:
+    """Sets up a persistent websocket."""
+
+    def __init__(self, endpoint_uri: URI, loop: asyncio.AbstractEventLoop, websocket_kwargs: Any) -> None:
+        """Initialize persistent websocket."""
+        self.ws: Optional[websockets.WebSocketClientProtocol] = None
+        self.endpoint_uri = endpoint_uri
+        self.loop = loop
+        self.websocket_kwargs = websocket_kwargs
+
+    async def __aenter__(self) -> websockets.WebSocketClientProtocol:
+        """Connect the websocket."""
+        if self.ws is None:
+            self.ws = await websockets.connect(uri=self.endpoint_uri, loop=self.loop, **self.websocket_kwargs)
+        return self.ws
+
+    async def __aexit__(self, exc_type: Type[BaseException], exc_val: BaseException, exc_tb: TracebackType) -> None:
+        """Close the websocket."""
+        if self.ws is not None and exc_val is not None:
+            try:
+                await self.ws.close()
+            except Exception:
+                pass
```

Comment:
Log exceptions on exit. Something like:

```python
except Exception as err:
    log.debug("Exception caught on exit", str(err))
```

You can pass the logger from provider to here.